### ISSUE TITLE
Alltoallv small to medium size messages desync, if used in gather-alike manner

### ISSUE THREAD

**Issue by DrHaraldBraun** [open-mpi/ompi#3255]
Hi,

repeated calls to Alltoallv exchanging small to medium size messages causes significant desynchronization of ranks, if used in a gather-alike manner. This leads to a major performance degradation in ompi 2.X.X as compared to ompi 1.X.X. All ompi 2.X Releases and release candidates exhibit this behavior.

The following was observed in a reproducer-run (see zip-archive) on 12 (2x18coreBDW_IB:FDR) nodes using 8 ranks per node -> commsize was 96 ranks.
Ratio>1 means Alltoallv+Barrier is faster than Alltoallv alone.
```
00: MPI_Alltoallv           sparse(gath) avMsgSize=8       B        : total 6.063e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=32      B        : total 5.471e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=128     B        : total 5.310e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=512     B        : total 5.129e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=2048    B        : total 2.597e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=8192    B        : total 2.349e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=32768   B        : total 3.779e+01 sec; 1000 calls
00: MPI_Alltoallv           sparse(gath) avMsgSize=131072  B        : total 3.291e+00 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=8       B        : total 6.385e-02 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=32      B        : total 6.465e-02 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=128     B        : total 6.655e-02 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=512     B        : total 7.415e-02 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=2048    B        : total 1.074e-01 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=8192    B        : total 2.553e-01 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=32768   B        : total 8.539e-01 sec; 1000 calls
00: MPI_Alltoallv+Barrier   sparse(gath) avMsgSize=131072  B        : total 3.057e+00 sec; 1000 calls
00: gather msg-size: 8          B -- ratio: 949.54    
00: gather msg-size: 32         B -- ratio: 846.34    
00: gather msg-size: 128        B -- ratio: 797.88    
00: gather msg-size: 512        B -- ratio: 691.64    
00: gather msg-size: 2048       B -- ratio: 241.83    
00: gather msg-size: 8192       B -- ratio: 92.02     
00: gather msg-size: 32768      B -- ratio: 44.26     
00: gather msg-size: 131072     B -- ratio: 1.08 
```
This behavior does not depend on the choice of pml (yalla, cm, or ob1), and seems to originate from the algorithms defined in coll.base or coll.basic via coll.tuned. 
A possible fix (see 'diff_patch_coll_sync_alltoallv.txt') is included in the tarball, which extends coll.sync to have an alltoallv method plus extra control variables ('coll_sync_barrier_after_alltoallv' and 'coll_sync_barrier_before_alltoallv' analoguous to the existing ones) . An alternative solution would be to exchange zero-Byte messages, like it used to be in ompi 1.X.X (?).

The reproducer (see source in 'alltoallv_gatheralike_test.c') is a gather-on-rank0-operation, however the actual usecase in the application is a gather-on-few-step of a two-phase-gather-operation (msgsize kB up to few 10 kB, without any synchronizing calls in the loop of gathers), but the observed behavior is qualitatively the same.

[Alltoallv_perf_fix.zip](https://github.com/open-mpi/ompi/files/879412/Alltoallv_perf_fix.zip)

**Comment by ggouaillardet** [comment]
Thanks for the patch
could you please revamp it (some files are not needed, ```Makefile.am``` is missing) and issue a PR vs the master branch ?
note you need to sign-off you patch so it can be accepted.

**Comment by jsquyres** [comment]
@DrHaraldBraun Yes, thanks for the patch!  As Gilles pointed out, can you please file this as a pull request?  That's our normal way of accepting / reviewing code contributions.  I have a few questions I'd like to ask about the patch.

Thank you!

**Comment by bosilca** [comment]
This is a lot of gymnastic for solving an issue that would better be solved at the level of the base algorithms. As indicated by @DrHaraldBraun bringing back the 0-bytes messages would automatically re-activate the synchronizing behavior of the algorithms, prevent the clogging of the PML with unexpected messages and solve the problem. In a much cleaner and consistent way.

**Comment by ggouaillardet** [comment]
@bosilca imho, ```MPI_Alltoallv()``` is faster without sending 0-bytes messages, just like ```MPI_Bcast()``` is faster without a ```MPI_Barrier()``` before or after.
this case is similar to a previous ticket in which ```MPI_Bcast()``` was involved in a loop.
at the end, bad things happen because of a specific pattern that leads to reduced overall performances (that includes hang) caused by too many unexpected messages.
```coll/sync``` was made to address this pattern, so i think it is right to extend it to handle this new pattern.

that being said, i am very pragmatic, and adding yet an other extra MCA parameter to send (or not) 0-bytes messages in the collectives is fine for me too.
if you think this is the way to go, i will be happy to make a new PR,
just tell me what should be the default value, i have no strong opinion on that.

**Comment by bosilca** [comment]
@ggouaillardet in what conditions MPI_Alltoallv is faster without sending 0 bytes messages (any scenario not in a hand-crafted benchmark) ? I am also pragmatic, and I think the added complexity of the sync (and the inherent weakness of the approach) make it unpalatable in this context.

As an example the proposed fix only work in the context of a single communicator. If the same pattern is repeated in duplicated communicators you will hit the same issue again. We can argue that this is a bad application, but it is legit from the standard perspective, and the proposed solution doesn't really fix the problem it just mend the only visible use case we know of right now.

**Comment by ggouaillardet** [comment]
@bosilca i simply meant that __no__ message is always faster than __one__ (even zero bytes) message.

so how to move forward ?
- __always__ send/recv zero bytes message ?
- makes it a MCA parameter (and with which default value) ?

**Comment by bosilca** [comment]
None of the proposed solutions are good. Pick the one that involves less efforts.

**Comment by DrHaraldBraun** [comment]
I chose the `coll/sync` approach for the very reason @ggouaillardet pointed out earlier. As this is likely a behavior that might only be seen for certain corner cases of `MPI_Alltoallv()`-usage, I intended to not alter behavior for applications, for which omitting zero byte messages might be advantageous. But I agree with @bosilca that this is quite some gymnastics to fix via `coll/sync`, where the cause of the issue is located in `coll/base`.
I also am pragmatic, so any solution is very welcome.

BTW: Does `MPI_Alltotallw()` as implemented in `coll/basic` suffer from the same issue, too, if used in a similar manner?

**Comment by jsquyres** [comment]
@ggouaillardet @bosilca Did #3285 solve this issue?

**Comment by DrHaraldBraun** [comment]
@jsquyres #3285 patched into ompi-master (539f71d0cca890a480d60efe7441a7423d6a7ee3) works fine for me (see timings in [a2av_ompi-master_patch3285.txt](https://github.com/open-mpi/ompi/files/916668/a2av_ompi-master_patch3285.txt) ). But the performance of the **not** patched master in the "Alltoallv+Barrier gatheralike" case is significantly better (compare with timings in [a2av_ompi-master.txt](https://github.com/open-mpi/ompi/files/916673/a2av_ompi-master.txt)).

And to answer my own question from above: Yes, `MPI_Alltoallw()` exhibits a similar behavior (see [a2aw_ompi-master.txt](https://github.com/open-mpi/ompi/files/916693/a2aw_ompi-master.txt)).
